### DISCUSSION TITLE
Skoleskema forkortninger

### DISCUSSION THREAD

**Discussion by PapaMikeJuliett** [scaarup/aula#117] [Q&A] [answered]
Hej

Jeg viser alle børnenes skoleskemaer i calendar card og vores skole bruger forkortelser, for eksempel: "MAT, LP".
Jeg vil gerne lave søg/erstat så der i stedet bliver vist: "Matematik, Lonnie Pedersen"

Er den funktionalitet allerede understøttet eller er der en måde at gøre det på?
Har forsøgt lidt forskelligt.
Calendar card understøtter ikke eventtemplate (ellers havde jeg "bare" gjort det)
Det er ikke lykkedes mig at lave en template calendar hvor jeg ændrer navnene
Har også forsøgt med atomic-calendar-revive card, men det lykkedes heller ikke.

Alle forslag er velkomne :)

Mvh
Palle

**Comment by scaarup** [comment]
Hej.
Det lader til at være lidt svært. Jeg kan ikke rigtig se nogen måde at hive calendar events op vha. templating.

**Reply by PapaMikeJuliett** [reply]
Ok.
Skal jeg oprette en idé til ny funktionalitet?  Jeg forestiller mig en eller anden configurering af integrationen / nogle linjer i config.yaml

**Comment by scaarup** [comment, marked as answer]
Jeg er lige kommet i tanke om en grim workaround du kan lave :). Skoleskemaet hives ned til en fil skoleskema.json hvert 10. minut. Hvis du kontinuerligt ligger og laver search/replace i den fil, så burde det blive reflekteret i HA.

**Reply by PapaMikeJuliett** [reply]
det lyder spændende :)  
Så de to "løsninger" jeg ser er:
1) stoppe AULA integrationen kl 6 om morgenen og 6:10 opdaterer jeg skoleskema.json.   Dog lidt uheldigt hvis der kommer en opdatering efter kl 6.
2) lave en automation der trigger på state-change på calendar og så opdatere skoleskema.json.   Dog lidt uheldigt at lav 144 disk writes på data der oftest er statisk

**Reply by dykandDK** [reply]
@PapaMikeJuliett 

Jeg ved ikke om du fik det løst, da det jo er en gammel tråd, men jeg har samme udfordring med forkortelser i mit skole skema og har løst det ved at lave en ny template sensor i HA, som opdateres hvert 30. minut ved at hente data fra kalenderen via den indbygge getEvents service.

Koden jeg bruger er denne:
```
template:
  - trigger:
      - trigger: time_pattern
        minutes: /30
    action:
      - action: calendar.get_events
        data:
          duration:
            hours: 120
            minutes: 0
            seconds: 0
          start_date_time: "{{ today_at('00:00') - timedelta(days=now().weekday()) }}"
        target:
          entity_id:
            - calendar.skoleskema_navn
        response_variable: calendars
      - variables:
          ev_list: |
            {###################################################}
            {# Makro, der ændrer forkortede navne på fag til   #}
            {# fulde fagbeskrivelser. F.eks. IDR til Idræt     #}
            {#                                                 #}
            {# Der kan tilføjes yderligere oversættelser efter #}
            {# behov, f.eks. for valgfag.                      #}
            {###################################################}
            {% macro fag_langt_navn(kortnavn) %}
            {{- kortnavn  | replace('BIL','Billedkunst')
                          | replace('BIO','Biologi')
                          | replace('DAN', 'Dansk')  
                          | replace('ENG','Engelsk')
                          | replace('FRA','Fransk')
                          | replace('FYS','Fysik/kemi')
                          | replace('GEO','Geografi')
                          | replace('HIS','Historie')
                          | replace('HDS','Håndværk og design')
                          | replace('IDR','Idræt')
                          | replace('KRI','Kristendomskundskab')
                          | replace('MAD','Madkundskab')
                          | replace('MAT','Matematik') 
                          | replace('MUS','Musik')
                          | replace('N/T','Natur/teknologi')
                          | replace('SAM','Samfundsfag')
                          | replace('TYSK','Tysk')
            -}}
            {% endmacro %}

            {###################################################}
            {# Tilføj evt. fag / tekster, som skal ignoreres   #}
            {# i ugeplanen, f.eks. dobbeltlektioner mv.        #}
            {###################################################}
            {%- set ignore_list = ['ÅP','Dobbl afk','PS'] -%}
            {###################################################}
            {# Tilføj forskydning i antal timer, hvis din Aula #}
            {# kalender ikke viser de korrekte timer           #}
            {###################################################}
            {%- set hour_offset = 2 -%}
            {% set ns = namespace(cal_events=[]) %}
            {%- for key, value in calendars.items() %}
              {%- for event in value.events %}
                {%- set fag = fag_langt_navn(event.summary.split(",")[0]) %}
                {%- if fag not in ignore_list %}
                  {%- set dato_tid = event.start %}
                  
                  {%- set start_tid = (as_datetime(event.start) + timedelta(hours=hour_offset)).strftime('%H:%M') %}
                  {%- set slut_tid = (as_datetime(event.end) + timedelta(hours=hour_offset)).strftime('%H:%M') %}
                  {%- set plan = {"dato_tid":dato_tid,"start":start_tid,"slut":slut_tid,"lektion":fag} %}
                  {%- set ns.cal_events = ns.cal_events + [plan] %}
                {%- endif %}  
              {%- endfor %}
            {%- endfor %}
            {{ ns.cal_events | sort(attribute='dato_tid') | list }}
    sensor:
      - name: aula_ugeplan2
        unique_id: aula_ugeplan2
        state: "{{ ev_list | count }}"
        attributes:
          lektioner: "{{ ev_list }}"
        icon: mdi:calendar
```

**Comment by scaarup** [comment]
Der ville jeg nok hellere lave en shell_command https://www.home-assistant.io/integrations/shell_command/, der fyrer dit script af - og så en automation i home assistant, der kører den hvert minut el.lign.

**Comment by scaarup** [comment]
Du kan jo teste det af manuelt med noget i retningen af:
`while true; do sed -i 's/MAT, LP/Matematik, Lonnie Pedersen/' skoleskema.json; sleep 60; done`

**Reply by PapaMikeJuliett** [reply]
det vil jeg prøve... dog med en eller anden "logik" på klokkeslaget.  ingen grund til at den gør det hele natten ;)
Mange tak for hjælpen :)

**Reply by PapaMikeJuliett** [reply]
Sidder og kigger i filen nu, og hele navnet er der faktisk, det er hele fagets navn dog ikke.

Her er hvad der er af info for hver felt i skoleskemaet (men det ved du måske nok allerede ;))
Jeg vil lave search / replace på \"teacherInitials\":\"LP\" og på \"title\":\"DAN\" og så se om det ikke slår igennem på skemaet.
Anyways, tak for hjælpen :)

`
{\"creatorInstProfileId\":null,

\"creatorProfileId\":null,
\"invitedGroups\":[{\"id\":972165,
\"name\":\"1A\",
\"shortName\":\"1\",
\"institutionCode\":\"705175\",
\"institutionName\":\"Gjern Skole\",
\"mainGroup\":true,
\"uniGroupType\":\"Hovedgruppe\",
\"isDeactivated\":false,
\"allowMembersToBeShown\":true}],
\"primaryResource\":{\"id\":30101,\"name\":\"64\"},
\"hasAttachments\":false,
\"createdDateTime\":\"2023-08-08T11:36:45+00:00\",
\"lesson\":{\"lessonId\":\"0bdd3a07-e384-44f9-b0ec-f73453670ac9\",
\"lessonStatus\":\"normal\",
\"participants\":[{\"teacherId\":1006523,
\"teacherName\":\"Lonnie Pedersen\",
\"teacherInitials\":\"LP\",
\"participantRole\":\"primaryTeacher\"}],
\"hasRelevantNote\":false},
\"timeSlot\":null,
\"vacationChildrenCountByDates\":null,
\"belongsToProfiles\":[3865124],
\"belongsToResources\":[],
\"oldAllDay\":null,
\"requiresNewAnswer\":false,
\"directlyRelated\":false,
\"responseDeadline\":null,
\"responseStatus\":null,
\"id\":416252018,
\"title\":\"DAN\",
\"allDay\":false,
\"startDateTime\":\"2023-09-12T06:10:00+00:00\",
\"endDateTime\":\"2023-09-12T06:50:00+00:00\",
\"oldEndDateTime\":null,
\"oldStartDateTime\":null,
\"responseRequired\":false,
\"private\":false,
\"type\":\"lesson\",
\"primaryResourceText\":null,
\"additionalResources\":[],
\"additionalResourceText\":null,
\"repeating\":null,
\"institutionCode\":null,
\"institutionName\":null,
\"addedToInstitutionCalendar\":false},
`

**Comment by scaarup** [comment]
Yes jeg ved det godt :)
Det er en beslutning jeg tog, kun at vise lærer initialer. Det er for at gøre det mere overskueligt og initialerne finder man hurgigt ud af. Dog hvis det er en vikar, er det et fuldt navn.

**Comment by goegsig** [comment]
Jeg har haft samme udfordring, jeg har lavet en fil der hedder mappings.py  som indeholder følgende 


```
SUBJECT_MAPPING = {
    "MAD": "🧑‍🍳  Madkundskab",
    "HIS": "🏛️ Historie", 
    "ENG": "🇬🇧 Engelsk",
    "DAN": "🇩🇰  Dansk",
    "MAT": "🧮 Matematik",
    "BIL": "🖼️ Billedkunst",
    "MB":  "📖 Morgenbånd",
    "N/T": "🤖 Natur/Teknik",
    "TYS": "🇩🇪 Tysk",
    "FRA": "🇫🇷 Fransk",
    "IDR": "🏃 Idræt",
    "KRI": "👼 Kristendom",
    "ELLE": "💬 Elfo",
    "PÆP": "🧩 Pædagogisk",
    "MUS": "🎸 Musik",
    "GEO": "🌍 Geografi",
    "HDS": "🔨 Håndværk og Design",
    "F. Valgf.": "🎯 Valgfag",
    "F/K": "🧪 Fysik/Kemi 🔥 💥 ⚗️ 🧬",
    "BIO": "🔬 🌱 Biologi"
}

TEACHER_MAPPING = {
    "CJ": "Charlotte ",
    "MFU": "Maja ", 
    "JD": "Jeppe ",
    "AHS": "Anne ",
    "HSC": "Henrik ",
    "AN": "Anna ",
    "REA": "Rikke ",
    "MIN": "Marlene ",
    "PW": "Per ",
    "FDH": "Frederik ",
    "LD": "Lise ",
    "SK": "Signe ",
    "J": "Rune ",
    "MS": "Mie ",
    "UB": "Ulla ",
    "GH": "Gitte ",
    "HJ": "Hanne ",
    "CSJ": "Cornelius ",
    "KL": "Kate ",
    "MN": "Maria ",
    "NN": "Nicolai ",
    "CØ": "Camilla "
}

```
Så i calendar.py tilføjer from .mappings import SUBJECT_MAPPING, TEACHER_MAPPING og parseCalendarLesson de to linjer der skifter lære og fag ud. Det virker super godt. 

```
from datetime import datetime, timedelta
import logging, time
from .const import DOMAIN
from homeassistant import config_entries, core
from .const import CONF_SCHOOLSCHEDULE
from homeassistant.components.calendar import (
    CalendarEntity,
    CalendarEvent,
)
from homeassistant.util import Throttle
## RLG ## 
from .mappings import SUBJECT_MAPPING, TEACHER_MAPPING
## RLG ## 


def parseCalendarLesson(lesson):
    summary = lesson["title"]
    # RLG Replace abbreviation with full form
    summary = SUBJECT_MAPPING.get(summary, summary)
    ##RLG
    
    start = datetime.strptime(lesson["startDateTime"], "%Y-%m-%dT%H:%M:%S%z")
    end = datetime.strptime(lesson["endDateTime"], "%Y-%m-%dT%H:%M:%S%z")
    location = (lesson.get("primaryResource", {}) or {}).get("name")
    vikar = 0
    for p in lesson["lesson"]["participants"]:
        if p["participantRole"] == "substituteTeacher":
            teacher = "VIKAR: " + p["teacherName"]
            vikar = 1
            break
    if vikar == 0:
        try:
            teacher = lesson["lesson"]["participants"][0]["teacherInitials"]
            # RLG Replace teacher initials with full name if available in the mapping
            teacher = TEACHER_MAPPING.get(teacher_initials, teacher_initials)
            #### RLG
        except:
            try:
                _LOGGER.debug("Lesson json dump" + str(lesson["lesson"]))
                teacher = lesson["lesson"]["participants"][0]["teacherName"]
            except:
                _LOGGER.debug(
                    "Could not find any teacher information for "
                    + summary
                    + " at "
                    + str(start)
                )
                teacher = ""
    lesson = CalendarEvent(
        summary=str(summary) + ", " + str(teacher),
        start=start,
        end=end,
        location=location,
    )
    return lesson

```
<img width="173" height="631" alt="billede" src="https://github.com/user-attachments/assets/97bb2619-234e-4d84-96d0-012443df1e90" />
